File: src/main/java/dev.toad/msg/Message.java
package dev.toad.msg;

import dev.toad.Debug;
import dev.toad.Eq;
import dev.toad.msg.option.Accept;
import dev.toad.msg.option.ContentFormat;
import dev.toad.msg.option.Host;
import dev.toad.msg.option.Path;
import dev.toad.msg.option.Query;
import java.net.InetSocketAddress;
import java.net.URI;
import java.net.URISyntaxException;
import java.util.List;
import java.util.Optional;

public interface Message extends Debug {
  public static dev.toad.msg.build.MessageNeeds.Destination builder() {
    return dev.toad.msg.build.Message.builder();
  }

  public Optional<InetSocketAddress> addr();

  public Id id();

  public Token token();

  public Code code();

  public Type type();

  public List<Option> options();

  public Payload payload();

  public dev.toad.msg.owned.Message toOwned();

  public byte[] toBytes();

  public default dev.toad.msg.build.Message buildCopy() {
    return dev.toad.msg.build.Message.copyOf(this);
  }

  public default dev.toad.msg.build.MessageNeeds.Code buildResponse() {
    return dev.toad.msg.build.Message.respondTo(this);
  }

  public default Optional<Option> getOption(long number) {
    return this.options().stream().filter(o -> o.number() == number).findAny();
  }

  public default Optional<Accept> getAccept() {
    return this.getOption(Accept.number).map(o -> new Accept(o));
  }

  public default Optional<ContentFormat> getContentFormat() {
    return this.getOption(ContentFormat.number).map(o -> new ContentFormat(o));
  }

  public default Optional<Path> getPath() {
    return this.getOption(Path.number).map(o -> new Path(o));
  }

  public default Optional<Host> getHost() {
    return this.getOption(Host.number).map(o -> new Host(o));
  }

  public default Optional<Query> getQuery() {
    return this.getOption(Query.number).map(o -> new Query(o));
  }

  public default URI uri() {
    int port = this.addr().map(a -> a.getPort()).orElse(5683);
    String scheme = port == 5684 ? "coaps" : "coap";
    String hostAddr =
      this.addr().map(a -> a.getAddress().getHostAddress()).orElse(null);
    String host =
      this.getHost()
        .map(h -> h.toString())
        .filter(h -> h != null && !h.trim().isEmpty())
        .orElse(hostAddr);
    String path =
      this.getPath()
        .map(p -> p.toString())
        .map(p -> p.startsWith("/") ? p : "/" + p)
        .orElse(null);
    String query = this.getQuery().map(q -> q.toString()).orElse(null);

    try {
      return new URI(
        scheme,
        /* userInfo */null,
        host,
        port,
        path,
        query,
        /* fragment */null
      );
    } catch (URISyntaxException e) {
      throw new RuntimeException(e);
    }
  }

  public static Eq<Message> eq() {
    return Eq.all(
      List.of(
        Eq.optional(Eq.socketAddress).contramap(m -> m.addr()),
        Eq.list(Option.eq()).contramap(m -> m.options()),
        Code.eq.contramap(m -> m.code()),
        Id.eq.contramap(m -> m.id()),
        Token.eq.contramap(m -> m.token()),
        Type.eq.contramap(m -> m.type()),
        Payload.eq.contramap(m -> m.payload())
      )
    );
  }

  public default boolean equals(Message m) {
    return Message.eq().test(this, m);
  }

  @Override
  public default String toDebugString() {
    return (
      this.type().toDebugString() +
      " " +
      this.code().toDebugString() +
      " " +
      this.uri().toString() +
      "\n  " +
      this.id().toDebugString() +
      " " +
      this.token().toDebugString() +
      this.options()
        .stream()
        .map(Debug::toDebugString)
        .reduce("", (b, a) -> b + "\n  " + a) +
      "\n" +
      (!this.payload().isEmpty() ? "\n" + this.payload().toDebugString() : "")
    );
  }
}

File: src/test/scala/Message.scala
import scala.jdk.CollectionConverters.*
import dev.toad.msg.*
import dev.toad.msg.option.ContentFormat
import dev.toad.msg.option.Host
import dev.toad.msg.option.Path
import dev.toad.msg.option.Query

class MessageBuilder extends munit.FunSuite {
  test("from(Message) copies another message") {
    val msg = dev.toad.msg.build.Message
      .builder()
      .uri("coap://localhost")
      .`type`(Type.NON)
      .code(Code.GET)
      .payload(Payload.json("[\"fart\"]"))
      .build

    assertEquals(msg, dev.toad.msg.build.Message.from(msg).build)
  }

  test("payload(Payload) sets content format to ContentFormat.JSON") {
    val msg = dev.toad.msg.build.Message
      .builder()
      .uri("coap://localhost")
      .`type`(Type.NON)
      .code(Code.GET)
      .payload(Payload.json("[\"fart\"]"))
      .build

    assertEquals(msg.getContentFormat.get, ContentFormat.JSON)
  }

  test("uri(String) uses system DNS to resolve host address") {
    val msg = dev.toad.msg.build.Message
      .builder()
      .uri("coap://localhost")
      .`type`(Type.NON)
      .code(Code.GET)
      .build

    assertEquals(msg.addr.get.getAddress.getHostAddress, "127.0.0.1");
  }

  test("uri(String) gets port from URI") {
    val msg = dev.toad.msg.build.Message
      .builder()
      .uri("coap://localhost:1234")
      .`type`(Type.NON)
      .code(Code.GET)
      .build

    assertEquals(msg.addr.get.getPort, 1234)
  }

  test("uri(String) gets port 5683 from scheme coap://") {
    val msg = dev.toad.msg.build.Message
      .builder()
      .uri("coap://localhost")
      .`type`(Type.NON)
      .code(Code.GET)
      .build

    assertEquals(msg.addr.get.getPort, 5683)
  }

  test("uri(String) gets port 5684 from scheme coaps://") {
    val msg = dev.toad.msg.build.Message
      .builder()
      .uri("coaps://localhost/cheese/gruyere?foo=bar&bingus")
      .`type`(Type.NON)
      .code(Code.GET)
      .build

    assertEquals(msg.addr.get.getPort, 5684)
  }

  test("uri(String) sets host to resolved host address from URI") {
    val msg = dev.toad.msg.build.Message
      .builder()
      .uri("coap://localhost/cheese/gruyere?foo=bar&bingus")
      .`type`(Type.NON)
      .code(Code.GET)
      .build

    assertEquals(msg.getHost.get.toString, "127.0.0.1")
  }

  test("uri(String) sets path to path section of uri") {
    val msg = dev.toad.msg.build.Message
      .builder()
      .uri("coap://localhost/cheese/gruyere?foo=bar&bingus")
      .`type`(Type.NON)
      .code(Code.GET)
      .build

    assertEquals(
      Seq.from(msg.getPath.get.segments.asScala),
      Seq("cheese", "gruyere")
    )
  }

  test("uri(String) sets query to query section of uri") {
    val msg = dev.toad.msg.build.Message
      .builder()
      .uri("coap://localhost/cheese/gruyere?foo=bar&bingus")
      .`type`(Type.NON)
      .code(Code.GET)
      .build

    val q = msg.getQuery.get.toMap.asScala
    assertEquals(Seq.from(q.get("foo").get.asScala), Seq(Query.Value("bar")))
    assertEquals(Seq.from(q.get("bingus").get.asScala), Seq(Query.Value.empty))
  }
}
